lib/utils/proxy/proxy_test.go

```diff
@@ -0,0 +1,146 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/lib/utils/socks"
+)
+
+// TestDialProxy tests the dialing mechanism of HTTP and SOCKS proxies.
+func TestDialProxy(t *testing.T) {
+	ctx := context.Background()
+	dest := "remote-ip:3080"
+
+	cases := []struct {
+		proxy      func(chan error, net.Listener)
+		scheme     string
+		assertDial require.ErrorAssertionFunc
+	}{
+		{
+			proxy:      serveSOCKSProxy,
+			scheme:     "socks5",
+			assertDial: require.NoError,
+		},
+		{
+			proxy:      serveHTTPProxy,
+			scheme:     "http",
+			assertDial: require.NoError,
+		},
+		{
+			proxy:      func(errChan chan error, l net.Listener) { close(errChan) },
+			scheme:     "unknown",
+			assertDial: require.Error,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.scheme, func(t *testing.T) {
+			errChan := make(chan error, 1)
+			l, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+
+			t.Cleanup(func() {
+				require.NoError(t, l.Close())
+				err := <-errChan
+				require.NoError(t, err)
+			})
+
+			go tc.proxy(errChan, l)
+
+			proxyURL, err := url.Parse(tc.scheme + "://" + l.Addr().String())
+			require.NoError(t, err)
+
+			conn, err := client.DialProxy(ctx, proxyURL, dest)
+			tc.assertDial(t, err)
+
+			if conn != nil {
+				result := make([]byte, len(dest))
+				_, err = io.ReadFull(conn, result)
+				require.NoError(t, err)
+				require.Equal(t, string(result), dest)
+			}
+		})
+	}
+}
+
+// serveSOCKSProxy starts a limited SOCKS proxy on the supplied listener.
+// It performs the SOCKS5 handshake then writes back the requested remote address.
+func serveSOCKSProxy(errChan chan error, l net.Listener) {
+	defer close(errChan)
+
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				errChan <- trace.Wrap(err)
+			}
+			return
+		}
+
+		go func(conn net.Conn) {
+			defer conn.Close()
+
+			write := func(msg string) {
+				if _, err := conn.Write([]byte(msg)); err != nil {
+					errChan <- trace.Wrap(err)
+					return
+				}
+			}
+
+			if remoteAddr, err := socks.Handshake(conn); err != nil {
+				write(err.Error())
+			} else {
+				write(remoteAddr)
+			}
+		}(conn)
+	}
+}
+
+// serveHTTPProxy starts a limited HTTP proxy on the supplied listener.
+// It performs the HTTP handshake then writes back the requested remote address.
+func serveHTTPProxy(errChan chan error, l net.Listener) {
+	defer func() {
+		close(errChan)
+	}()
```

Comment:
nit: defer close(errChan)  for consistency with https://github.com/gravitational/teleport/pull/17976/files#diff-ee03621033de1f1c966e77b73706eb4db1e782434e2710e94341a4b30f5002acR95